text/1015-full-text-search.rst

```diff
@@ -0,0 +1,368 @@
+introduces ``ftsearch`` scalar type. It is meant to be analogous to
+``tsquery`` and it should be capable or representing common queries regardless
+of the specif FTS backend (Postgres, Elasticsearch, Lucene, etc.).
+
+The searching functionality can be split along two axis:
+1) what FTS operates on (``str`` vs ``Object``)
+2) what FTS returns (``bool`` vs some ranked results)
+
+The task of testing whether a document matches an FTS query would be performed
+by the following built-in functions::
+
+  function fts_test(doc: str, query: ftsearch) -> bool
+  function fts_test(doc: Object, query: ftsearch) -> bool
+
+The input can be either a ``str`` or an entire ``Object``. The object type
+must have some ``str`` properties, however. Trying to perform FTS on a type
+without ``str`` properties is an error. This function is generally intended to
+be used in ``filter`` clauses in cases where there's no need for hinting what
+matched, but the results are simple enough to be self-explanatory.
+
+The ranked results need to be rich objects with several relevant properties,
+akin to free objects used for returning ``group`` results. The difference is
+that we cannot use a ``FreeObject`` as a return type of a function, we need a
+specialized type for that: ``FTSResult``. This type is defined as follows::
+
+  type FTSResult {
+    property match -> str;
+    link objectMatch -> Object;
+
+    property rank -> float64;
+    property highlights -> array<str>;
+  }
+
+Depending on whether a ``str`` or ``Object`` was used as input, the ``match``
+or ``objectMatch`` will be populated, respectively, while the unused "match"
+will be ``{}``.
+
+If ranking was requested, then the value will be written in ``rank``.
+
+If highlighting the matched text was requested, then the appropriate value
+will be written in ``highlights`` array. The array is needed in case the
+matches are highlighted separately (e.g. only the matched words are returned,
+omitting the surrounding text).
+
+These enhanced FTS matches would be performed by the following built-in
+functions::
+
+  function fts_match(
+    doc: str,
+    query: ftsearch,
+    rank_opts: optional str = 'default',
+    highlight_opts: optional str = {}
+  ) -> FTSResult
+
+  function fts_match(
+    doc: Object,
+    query: ftsearch,
+    rank_opts: str = 'default',
+    highlight_opts: optional str = {}
+  ) -> FTSResult
+
+The ``rank_opts`` and ``highlight_opts`` can potentially be backend-specific,
+however either one of them can take the value of ``{}`` with the meaning that
+ranking/highlighting of the results is not necessary. Both of them should also
+always accept ``default`` as a valid argument indicating that
+ranking/highlighting of the results should be done in some basic manner
+(typically using some faster method).
+
+We also want some way of telling when one ``ftsearch`` value is encountered as
+part of another::
+
+  function contains(haystack: ftsearch, needle: ftsearch) -> bool
```

Comment:
Hm, can we use `contains` instead of `fts_test`?

Reply:
I was thinking back to regexp test/match duality here. I'm not hating `contians` here, though. If there is a vote or two more for it, we could change that.

Reply:
Is `haystack: ftseach` a typo? Or what the idea behind checking whether a search term contains another search term?